include/openssl/asn1.h

```diff
@@ -248,11 +248,17 @@ typedef struct ASN1_VALUE_st ASN1_VALUE;
         type *name##_new(void); \
         void name##_free(type *a);
 
-# define DECLARE_ASN1_PRINT_FUNCTION(stname) \
+# define DECLARE_ASN1_DUP_FUNCTION(type) \
```

Comment:
The benefits I had in mind introducing `DECLARE_ASN1_DUP_FUNCTION` are 
* avoid redundancy when declaring `_dup` functins, which otherwise mention (partly rather long) type names/expressions three times. For instance
```
ASN1_OCTET_STRING *ASN1_OCTET_STRING_dup(const ASN1_OCTET_STRING *a);
```
reduces to
```
DECLARE_ASN1_DUP_FUNCTION(ASN1_OCTET_STRING)
```
* uniformity declaring  `_dup` functions

Reply:
Do all ASN.1 types have/offer a  `_dup` function? Then `DECLARE_ASN1_DUP_FUNCTION` could be merged into `DECLARE_ASN1_FUNCTIONS`.

Reply:
Only very few ASN.1 types offer dup functionality: `ASN1_INTEGER`, `ASN1_STRING` and `ASN1_OCTET_STRING`.
While I'm all for completeness, I'm also pragmatic, so I'm unsure this is necessary, we can live with the exceptional unmacrod declaration of the few functions that actually need declaring.
If we had littered all ASN.1 types with dup functionality, it would have been a different matter (and a DUP macro would have appeared long ago)

Reply:
@levitte, see my answer below, where you duplicated your comment.

Reply:
A further motivation I had for introducing `DECLARE_ASN1_DUP_FUNCTION` that I forgot to mention earlier (I've just updated my post above) is that 
there is already an `IMPLEMENT_ASN1_DUP_FUNCTION ` macro.

Reply:
> Only very few ASN.1 types offer dup functionality: ASN1_INTEGER, ASN1_STRING and ASN1_OCTET_STRING.

Are you sure?
I'm not an expert on ASN.1 and its OpenSSL implementation, but for instance
`grep IMPLEMENT_ASN1_DUP_FUNCTION crypto/*/*.c`
yields 22 hits 
(which Interestingly does not include `ASN1_INTEGER`, `ASN1_STRING`, and `ASN1_OCTET_STRING`):
```
crypto/asn1/x_algor.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_ALGOR)
crypto/ocsp/ocsp_lib.c:IMPLEMENT_ASN1_DUP_FUNCTION(OCSP_CERTID)
crypto/pkcs7/pk7_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(PKCS7)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(TS_MSG_IMPRINT)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(TS_REQ)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(TS_ACCURACY)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(TS_TST_INFO)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(TS_STATUS_INFO)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(TS_RESP)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(ESS_ISSUER_SERIAL)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(ESS_CERT_ID)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(ESS_SIGNING_CERT)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(ESS_CERT_ID_V2)
crypto/ts/ts_asn1.c:IMPLEMENT_ASN1_DUP_FUNCTION(ESS_SIGNING_CERT_V2)
crypto/x509/x_attrib.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_ATTRIBUTE)
crypto/x509/x_crl.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_REVOKED)
crypto/x509/x_crl.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_CRL)
crypto/x509/x_exten.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_EXTENSION)
crypto/x509/x_name.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_NAME_ENTRY)
crypto/x509/x_name.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_NAME)
crypto/x509/x_req.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509_REQ)
crypto/x509/x_x509.c:IMPLEMENT_ASN1_DUP_FUNCTION(X509)
```

Reply:
I just (re)discovered the same thing myself.  So yeah, the existence of `IMPLEMENT_ASN1_DUP_FUNCTION` is a strong reason to have a corresponding `DECLARE_ASN1_DUP_FUNCTION`.  So ok.

Reply:
Looking at `include/openssl/asn1.h`, I wonder if the declarations shouldn't be remade to generate static inline functions.  That would avoid the need for the IMPLEMENT macros...  buuut I think that should be done in a different PR.